### PR TITLE
[IMP] core: allow some users to bypass the sanitize of HTML field

### PR DESCRIPTION
**Pull request by rdeodoo**
[IMP] core: allow some users to bypass the sanitize of HTML field
Add the possibility to flag a HTML field as `sanitize_overridable`.
The sanitizer will then be bypassed if the user doing the operation is
part of the new group `base.group_sanitize_override`.
The "Settings" users are part of that new group.

If such a user wrote some HTML that would have normally been removed by
the sanitizer, then users without the right to bypass the sanitizer
won't be able to write on that field anymore.
Otherwise, it would sanitize the previously written data.
Such cases are detected, and the modification prevented by the system,
which will warn the user about it.

For instance, with a `field.html(sanitize_overridable=True)`: one being
part of `group_sanitize_override` could write `<script></script>`.
Then, someone not part of the group trying to add an element inside that
field like appending a `<p/>` -> `<script></script><p>New Content</p>`
would not be able to because it would go through the sanitizer and
ultimately, removing part of the original value:
`<p>New Content</p>` (`<script></script>` would be removed).

== Real use case ==
In the website builder, there is 2 editor rights:
- group_website_publisher: restricted editor
- group_website_designer: editor & designer

The designer editor can edit pages and views, while the restricted
editor can't do anything unless he is part of other groups.
In edit mode, the restricted editor will only be able to edit fields of
record he has access to.
For instance, being a sales manager allows you to edit a product
description on the website.
Being an event manager -> edit event. Slide manager -> Slides etc.
As those restricted editor are able to edit those fields, they are
(almost) always sanitized to prevent them to introduce malicious code.

Since those fields are sanitized, even admins / designer editor are not
able to fully use the website builder in such fields.

Some clients don't really care about that sanitation, they'd prefer to
avoid it as they trust their manager and would prefer to have the full
builder capability instead.
This is typically the case in small project (butcher, hairdresser,
reseller etc) and in SMEs.

With the new `sanitize_overridable` feature, they will be able to do
that, as the "Designer & Editor" group now also receive the group
`base.group_sanitize_override` (done in next commit).

### CODE REVIEW

**Review by qsm-odoo — COMMENTED** [review]
Hello @odoo/rd-framework-py 

Could you validate the approach here please? :) (the first commit: `[IMP] core: allow defined groups to bypass the sanitize of HTML field`, the rest (UX) will be further improved when we get the green light from you :+1:).

I think the commit message RDE made explains the need and new feature very well. One missing information for you: this is a combined idea of FP and AL :wink: 

Thanks in advance :+1:

File: odoo/fields.py
Comment (Julien00859) [inline odoo/fields.py:1902]: if (not initial_value_sanitized
        or get_parsed(original_value) != get_parsed(initial_value_sanitized)):

> The preferred way of wrapping long lines is by using Python’s implied line continuation inside parentheses, brackets and braces. Long lines can be broken over multiple lines by wrapping expressions in parentheses. These should be used in preference to using a backslash for line continuation.

> Use `is not` operator rather than `not ... is`. While both expressions are functionally identical, the former is more readable and preferred.

<https://peps.python.org/pep-0008/>

File: odoo/fields.py
Comment (Julien00859) [inline odoo/fields.py:1889]: We are trying to favorise `isinstance(..., collections.abc.Iterable)` over `isinstance(..., list)`. The former allows using other structures such as tuples, sets or UserLists. The latter forces using a list. The current state of `fields.py` is that there are still *lots of* `isinstance(..., list)` so feel free to dismiss this comment. In the docstring too.

**Review by Julien00859 — COMMENTED** [review]
Leaving apart the security considerations, the implementation seems fine to me.

**Review by xmo-odoo — COMMENTED** [review]
Not judging the merits of the idea yet, but the interface (UX) is really bad, the way it reads tells the user the exact opposite of what it does.

File: odoo/fields.py
Comment (rdeodoo) [reply odoo/fields.py:1889]: Changed :+1:

File: odoo/fields.py
Comment (rdeodoo) [reply odoo/fields.py:1902]: :+1: 
I always hesitate on this one. Thanks.
I went for the same style as the multi line conditions in this file
```py
                if (
                    not initial_value_sanitized
                    or get_parsed(original_value) != get_parsed(initial_value_sanitized)
                ):
```

**Comment by rdeodoo** [comment]
Thanks @Julien00859 suggestions applied.

@xmo-odoo 
> the interface (UX) is really bad, the way it reads tells the user the exact opposite of what it does.

The website implementation part is, as commit suggest, a WIP to illustrate the use case for now:
```
[WIP] web_editor, website: warn restricted editor about forbidden change
See previous commit for more details.
This commit is just about implementing a basic first solution for the
website builder to easily be able to figure and try the use case / need
of that behavior.

Both the final functionnal implementation and the commit message will be
rewritten.
```
 So I guess you are talking about the message raised in `fields.py` (`Someone with escalated rights previously modified this field (%s %s), you are therefore not able to modify it yourself.`)?

This message is probably to be rewritten for a potential final version, maybe something shorted, clearer and less technical (`escalated rights` / `field`). I'll leave that final decision to the POs and/or the framework team anyway.

But what exactly do you mean by `the exact opposite of what it does`?
If a change is to be made on an `HTML` field, the original value (before the change is actually made) is checked to see if that existing part would be sanitized by the restricted user. If so, it would mean that the current restricted user would alter the original value due to the sanitation, meaning the original value was set by someone with escalated rights (or a potential default value).

**Comment by xmo-odoo** [comment]
> So I guess you are talking about the message raised in `fields.py` (`Someone with escalated rights previously modified this field (%s %s), you are therefore not able to modify it yourself.`)?

No, I'm talking about `sanitize=list`, that gives the impression that sanitization applies only to those groups rather than the opposite. I like that it's an allowlist (rather than a denylist), but it doesn't *read* that way when looking at the code.

Maybe `sanitize` should be a dict with a bunch of configuration knobs, and the allowlist could be a `except` key or something along those lines.

**Comment by qsm-odoo** [comment]
> No, I'm talking about sanitize=list, that gives the impression that sanitization applies only to those groups rather than the opposite. I like that it's an allowlist (rather than a denylist), but it doesn't read that way when looking at the code.

Indeed, we actually said the same thing. Something like it could be another value like `sanitize_bypass` or something like that so the API is clearer. We'll let you decide :+1:

**Comment by rdeodoo** [comment]
Note that the idea was to not introduce another new `sanitize` related attribute but use the existing one, despite the "confusion".

Moving all the sanitize attributes in a dict is probably cleaner :+1: (would need a bigger refactoring)
Let us know if I should change the pr in that direction.

**Comment by mart-e** [comment]
Is this really a good idea? I am not fan of adding new ways to bypass the sanitation.

> Some clients don't really care about that sanitation

No, they don't _understand_ sanitation and this is why we impose it to them (the same way you don't ask the opinion of people saying they don't care about privacy). It's not just a "you can't embed that youtube video" feature, it's a "you can no longer trust the content of this field anywhere". We try very hard (not always successufully, I admit) to maintain the `sanitize=False` fields to a strict minimum of restricted models with clear acess control.
This could be a good news if it allows use to convert some `sanize=False -> sanitize=groups`. But this is the opposit in this PR (and I guess, in future PRs too), we are disabling sanitation on several models. It's even more difficult to know if you can trust the content of a field with this attribute.
Should we have to add `with_user` to th ci/security as it's now becoming a way to circumvent sanitation?

**Comment by qsm-odoo** [comment]
@mart-e 

> No, they don't understand sanitation and this is why we impose it to them

We all know that of course, but here we are talking about giving that right to not sanitize to users who can already add non-sanitized code to most parts of the website (all views). But everything you said make sense of course. Indeed, that would decrease security. But I don't think there is a way to not decrease security and allow that important feature :cry:

**Comment by fpodoo** [comment]
> No, they don't _understand_ sanitation and this is why we impose it to them

You might misunderstand the issue.

In master, sanitation is defined **per type of content**: a web page is not sanitized, a product page is. It leads to weird trade-offs security vs feature, that's why a blog post is not sanitized.

This PR will allow to sanitize more fields (e.g. blog post) by moving the security to the user group, in addition to the type of content. Both `group_website_publisher` and `group_website_designer` can edit a blog post, but only publisher can do unsanitized content.

Both security & features will win:
- we'll be able to sanitize blog post, for untrusted users: more security
- we'll allow trusted users to put any building block in a product description: more features

Moving the security to the user group, instead of the type of content seems a better design.

Extra Note: Website content and blog posts are currently not sanitized (ir.ui.view). There are plenty of reasons why website pages can not be sanitized (embed video, JS, ...). So, no, we can not impose sanitation to all users, we need to trust some users (the group `group_website_designer`)

**Comment by mart-e** [comment]
So to sum up, the rationale is that a user that can write unsanitised HTML (to include videos, JS snippets,...) in website pages, he should also be able to do so in blog posts or product description. 
I discussed with the others from the team and we all agree on the intention, it's more consistent :+1: 
However, with this implementation, this is opening a dangerous hole in the sanitation, it is critical we ensure this won't introduce side effects or be abused in the future.

One way to avoid this, could be to work the other way around. We could create a specific group that is allowed to bypass the sanitation and simply mark the fields and "overridable" for its members.
Suggestion:
1. create a field flag `sanitze_overridable=True` (feel free to suggest better names)
2. create a `base.group_sanitize_override` group, add it to the `implied_ids` or `base.group_system` and `website.group_website_designer`.
This is similar to what we did with email templates and the `mail.group_mail_template_editor`.

Not so much security related but regarding blocking modifications for untrusted users, I think it's better to let any user modify it:
- Technically, it's quite difficult to work with the "previous value" (cf all the issues we have with cache and dirty detection) so this may lead to false negatives and complexe code.
- For the user experience, I find it worse to fully block the edition that to risk losing some content. Let say the only manager is away and you absolutly need to modify a product description (e.g. refering to outdated promotion,...), it's better to lose some rich content than being unable to modify it. This is also closer to what currently happens (losing content when pressing the "Save" button) so I guess user will be able to live with that. In the case of the small company where everybody is trusted and has admin priviledges (which is the target of this PR as I understand), this won't change anything.

**Comment by rdeodoo** [comment]
Thanks for the sum up, glad we found a compromise.
I'll adapt the PR with technical changes suggested.

About the "let user erase part of the DOM" > "prevent and warn", I don't really agree, imagine someone bypassed the sanitation and made a very nice page with snippet code (3D product presentation snippet) etc.
Then a sales manager is trying to fix a typo he found, he will break the page.

**Comment by qsm-odoo** [comment]
Good idea for the dedicated group but for the rest, same opinion as @rdeodoo 

> Let say the only manager is away and you absolutely need to modify a product description

We can always find cases for everything if we phrase sentences like that :wink: I can say the same sentence to prove the other point: "Let say the only manager is away and you absolutely need to modify a product description... the system prevents you doing it and breaking the important embed code a manager added without any possibility to restore it while he is gone".

> This is also closer to what currently happens (losing content when pressing the "Save" button) so I guess user will be able to live with that.

This happened before, this is not supposed to anymore. As discussed with ODO before, we implemented something to prevent users to add things that will be removed by sanitization. Of course there is always a way if they use some custom snippets or whatever but...

> it's quite difficult to work with the "previous value" (cf all the issues we have with cache and dirty detection)

It's probably not the same things in that case? We don't try to compare the "previous value" with a new one, we compare the "previous value" with itself to determine if it has already been sanitized or not -> if sanitized, then we can add anything if correct rights. If not sanitized, then we must preserve the non-sanitization and thus block users without access rights.

> this may lead to false negatives and complexe code.

Maybe I don't see it :shrug: In any case, the current code is not complex though and if it leads to false positive/negative, it's probably not critical (we would just lose content like you seem to consider acceptable or block the user asking him to get a manager (which I find acceptable as explained above)). Of course if it's 50% false positive/negative, it's not the same story :grimacing: 

In any case, I can live with any decision regarding this task really :wink: In any case, it will be a good progress for the builder :+1:

**Comment by mart-e** [comment]
The second part was more subjective, I am not a huge fan of error popups but your arguments make sense as well. I guess it's for the website PO to decide this one :)

**Comment by rdeodoo** [comment]
@mart-e I pushed the implementation you suggested in a `WIP` commit on top of the PR. It is adapting all 3 previous commits.
I'll move that new commit inside the 3 previous ones once that technical change is validated on your side :+1:

**Comment by pierre-pde** [comment]
Hi @rdeodoo , @qsm-odoo  & @mart-e , 
I confirm that I prefer that restricted editors are *not* able to update field that would loose content if they would edit it.
We will have to make it clear, in the error message pop-up, that it's because an editor has added restricted content before them.

**Comment by rdeodoo** [comment]
The PR has been fine tuned with @mart-e suggestion:
- It now uses `sanitize_overridable` boolean attribute instead of allowing a collection of groups to be passed to the `sanitize` attribute
- That new attribute is linked to the `base.group_sanitize_override` new group which "Settings" group and "Designer & Editor" group receive automatically.

File: odoo/fields.py
Comment (rdeodoo) [inline odoo/fields.py:1890]: The `not record.env.registry._init` part is needed as we still want to `html_sanitize` the content during the XML import.
Indeed, some records' fields' values are ("wrongly"?) set as `type="xml"` while they are pure HTML value inside an HTML field. Without going through the sanitizer, the value will be prefixed/wrapped into

```xml
<?xml version="1.0"?>                                                                           +
<data><!-- ACTUAL VALUE --></data>
```
(wrapped only if multiple root nodes)

Thus, this PR would change the value added in DB for such cases.

It would then mean that for those (very rare) cases, if a restricted editor would try to modify it, it would directly be catch as unsafe modification due to the `<data/>` wrapper.

Note that I made a PR to "fix" those occurrences: https://github.com/odoo/odoo/pull/98239

-------------

Initial TMP fix was to also check for similarity with the `<data/>` wrapper, but:
1. It was doing some extra operation / slowing down the process for no reason
2. For those rare cases, it was inserting values in DB differently than before the PR (wrapping into `<data/>`) 

```py
    and get_parsed(original_value) != get_parsed('<data>%s</data>' % initial_value_sanitized)
```

**Comment by rdeodoo** [comment]
Hello @odoo/rd-security 
Please see https://github.com/odoo/odoo/pull/97398#issuecomment-1210687733, the PR is ready, any review would be appreciated :+1:

**Comment by xmo-odoo** [comment]
@robodoo override=ci/security

File: addons/web_editor/models/ir_qweb_fields.py
Comment (qsm-odoo) [inline addons/web_editor/models/ir_qweb_fields.py:392]: This should actually be part of the first commit, only the part `not has_group('sanitize_override')` should be part of this commit. But after the first commit, it's already true that a sanitize_overridable field for someone who *has* the sanitize_override group should not be marked with `data-oe-sanitize`.

Nitpicking I guess, especially if we `rebase-merge` but even in that case... :) (btw would you prefer rebase-merge or rebase-ff ?)

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:219]: I know there has been debate about this and as I said, I preferred this version but maybe "area" instead of "field", maybe a little more user friendly?

Also, I still think we should find better that an alert on every click but this PR can move on without something fancy for this IMO.

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
Some details then I'll test and merge if no objection :+1:

**Review by rdeodoo — COMMENTED** [review]
> btw would you prefer rebase-merge or rebase-ff ?

`rebase-ff` is perfectly fine for me, but if you prefer a `rebase-merge` it also makes sens and I'll be fine with it too.

File: addons/web_editor/models/ir_qweb_fields.py
Comment (qsm-odoo) [reply addons/web_editor/models/ir_qweb_fields.py:392]: Re-read what I said ^^ I am not saying the part in "not has_group" should be in the first commit, I am saying the exact opposite :grimacing: 

> after the first commit, it's already true that a sanitize_overridable field **for someone who has the sanitize_override group** should not be marked with data-oe-sanitize

The code I comment is:
```
if sanitize_overridable
     if not has_group
          do something
     else
          do nothing
else
     add sanitize attribute
```

The kinda-hidden "do nothing" part, should be part of the first commit, the do something part should stay in the commit it is currently in.

Comment (rdeodoo) [reply addons/web_editor/models/ir_qweb_fields.py:392]: > which is what you did in the end

Well, if despite the misunderstanding, the code was adapted as planed, the change certainly made sense :+1: 

Note that about the PR, I believe everything from everyone has been processed so far.

**Review by qsm-odoo — APPROVED** [review]
@robodoo rebase-ff r+

**Pull request closed without merging**
